soundcard/pulseaudio.py

```diff
@@ -333,26 +332,17 @@ class _Recorder(_Stream):
 
     """
 
+    def __init__(self, id, samplerate, channels, blocksize=None,
+                 name='outputstream', fix_blocksize=False):
+        super(_Recorder, self).__init__(id, samplerate, channels, blocksize, name)
+        self._pending_chunk = numpy.zeros((0, ))
+
     def _connect_stream(self, bufattr):
         self._pulse._pa_stream_connect_record(self.stream, self._id.encode(), bufattr, _pa.PA_STREAM_ADJUST_LATENCY)
 
-    def record(self, numframes):
-        """Record some audio data.
-
-        The data will be returned as a `frames × channels` float32
-        numpy array.
-
-        This function will wait until `numframes` frames have been
-        recorded. However, the audio backend holds the final authority
-        over how much audio data can be read at a time, so the
-        returned amount of data will often be slightly larger than
-        what was requested. The amount of buffering can be controlled
-        through the blocksize of the recorder object.
-
-        """
-
-        captured_frames = 0
+    def _record_chunk(self, numframes):
```

Comment:
This function should not take `numframes` as an argument, and instead only return exactly zero or one chunk. This will simplify the function greatly, and get rid of the loop.